server/grpc_test.go

```diff
@@ -464,10 +465,36 @@ func TestGrpcByteStreamDeadline(t *testing.T) {
 	}
 
 	time.Sleep(time.Millisecond)
+	ctx, cancel = context.WithTimeout(context.Background(), time.Millisecond*500)
```

Comment:
The code from this point onwards feels like it could be a separate test.

But maybe this is already covered by `TestGrpcByteStream`, which does a bytestream Write followed by Read- I guess that's effectively the same, just without a timeout. Maybe we should add a timeout there instead?

Reply:
This test shows that a canceled upload doesn't leave the cache in a bad state and prevent future uploads from being discarded.

This first portion of this test actually passes without the fix. We won't fail any assertions until trying to read the file from the cache, after the second successful write (which was actually discarded) and finding it is not there.

Reply:
Ah, I see.